### ISSUE TITLE
Functionality to identify which input parameters trigger a given classification (time-series data)

### ISSUE THREAD

**Issue by daanvdn** [deeplearning4j/deeplearning4j#3568]
I am training a multilabel classifier which takes time-series data as input and predicts for a given input which set of labels should be assigned to that input.

As can be seen from the following (hypothetical) examples the number of features is stable between cases, whereas the number of time steps can vary across cases:

```
labeled_example_1 => labels: rain, snow, clouds  
featureId:              0   1   2   3   4   5   6   7   8   9
featureValue_timeStep_1 1.5 0.8 11  1.8 20  100 6   88  0   100
featureValue_timeStep_2 1.5 0.6 3   1.2 800 99  5   0.1 5   105
featureValue_timeStep_3 1.5 1.8 0   0.8 2.0 110 1.0 0.8 6   200
featureValue_timeStep_4 0.5 8.8 0.1 1.8 20  500 5   68  0   0.05
featureValue_timeStep_5 1.6 3.7 12  0.8 20  122 8   8.8 2   888
featureValue_timeStep_6 9.5 5.8 150 2.8 20  0.6 100 880 6   300

labeled_example_2 => labels: rain, sunny  
featureId:              0   1   2   3   4   5   6   7   8   9
featureValue_timeStep_1 2.5 0.8 20  1.8 100 100 6   98  6   100
featureValue_timeStep_2 1.5 3.6 3   0.2 800 88  1   5.1 5   105

labeled_example_3 => labels: storm, tornado
featureId:              0   1   2   3   4   5   6   7   8   9
featureValue_timeStep_1 2.5 0.8 20  1.8 100 100 6   98  6   100
featureValue_timeStep_2 1.5 3.6 3   0.2 800 88  1   5.1 5   105
featureValue_timeStep_3 1.5 2.8 0   1.8 1.0 210 1.0 0.9 6   600
```

The number of features in real-world cases is obviously much larger -- around 7000 -- and also the max. amount of time steps is much bigger (between 1 and 160).

My classification problem is such that each example can be classified as belonging to one or more classes (in total there are about 40 different classes). To achieve this I've configured a MultiLayerConfiguration along the following lines:

```java
new NeuralNetConfiguration.Builder().optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
				.iterations(1)
				.learningRate(0.002)
				.rmsDecay(0.95)
				.seed(12345)
				.regularization(true)
				.l2(0.001)
				.weightInit(WeightInit.XAVIER)
				.updater(Updater.RMSPROP)
				.list()
				.layer(0, new DenseLayer.Builder().nIn(7000).nOut(2000).activation("tanh").build())
				.layer(1, new GravesLSTM.Builder().nIn(2000).nOut(300).activation("tanh").build())
				.layer(2,
						new RnnOutputLayer.Builder(LossFunctions.LossFunction.XENT).activation("sigmoid")
								.nIn(300)
								.nOut(40)
								.build())
				.backpropType(BackpropType.Standard)
				.pretrain(false)
				.backprop(true)
				.setInputType(InputType.recurrent(nIn))
				.build();
```

Once I have trained this model I want to use it in production to achieve 2 distinct goals (cf. example below):
(1) classify an unseen case as belonging to 1 or more of the output classes
(2) for a given classification of a given case __determine exactly which input features trigger the predicted class(es)__.

unseen_example => predicted labels: __tornado, clouds__
featureId:              0   1   2   3   4   5   6   7   8   9
featureValue_timeStep_1 2.5 0.8 20  __1.8__ 100 100 6   98  6   100
featureValue_timeStep_2 1.5 3.__6__ 3   0.2 800 88  __1__   5.1 5   105
featureValue_timeStep_3 __1.5__ 2.8 __0__   1.8 1.0 210 1.0 0.9 6   __600__

Whereas (1) is supported out of the box by the dl4j api, it's goal (2) where I am stuck.

A while ago I have been discussing this with @turambar (see: (https://gitter.im/deeplearning4j/deeplearning4j?at=58af122e872fc8ce62753468)) and according to him this functionality -- "called model interrogation via gradient descent on inputs" -- has already been considered to be put on the roadmap (apparently @AlexDBlack is also involved in this discussion)

My questions with regard to this are as follows:
- are there already any concrete plans to add this functionality?
- if not, is it feasible for somebody who is not a DL expert to implement it? In other words: is it an engineering problem or more a mathematical / data science problem? If it is more an engineering problem than anything else, I would be very happy to work on it. However, I'd still need some pointers on where to start...

Cheers

daanvdn

PS: possible duplicate of (https://github.com/deeplearning4j/deeplearning4j/issues/2866)

**Comment by AlexDBlack** [comment]
What you're asking for here is known as model interpretability. There are a number of different approaches to this, with different pros and cons. Unfortunately we have not yet formally implemented any of the approaches in DL4J.

When dealing with timeseries, the following (non-exhaustive list of) approaches are available:

- RNN attention models – not available in DL4J yet
- heat/saliency maps via occlusion or replacement (Zeiler 2014, Zingraf 2016)  – easy enough to do, without modifying or changing any of the networks/layers
- heat/saliency maps via gradients (Simonyan 2013) – also easier
- layer-wise relevance propagation (Bach 2015) - hard to implement
- back propagation of activation differences (Shrikumar 2016) - hard?



Matthew D Zeiler and Rob Fergus. Visualizing and understanding convolutional networks. In European Conference on Computer Vision, pages 818{833. Springer, 2014.
Luisa M Zintgraf, Taco S Cohen, and MaxWelling. A new method to visualize deep neural networks. arXiv preprint arXiv:1603.02518, 2016.
Karen Simonyan, Andrea Vedaldi, and Andrew Zisserman. Deep inside convolutional networks:
Visualising image classification models and saliency maps. arXiv preprint arXiv:1312.6034, 2013.
Sebastian Bach, Alexander Binder, Gregoire Montavon, Frederick Klauschen, Klaus-Robert Muller, and Wojciech Samek. On pixel-wise explanations for non-linear classifier decisions by layer-wise relevance propagation. PloS one, 10(7):e0130140, 2015.
Avanti Shrikumar, Peyton Greenside, Anna Shcherbina, and Anshul Kundaje. Not just a black box: Learning important features through propagating activation differences. arXiv preprint
arXiv:1605.01713, 2016.



With regard to heat/saliency maps via gradients, from what I recall from reading the paper (correct me if I'm wrong - I haven't gone back to check) they simply utilise the gradients of the loss function with respect to the inputs – that is, dL/dIn. How to get that? Set your inputs and labels, call computeGradientAndScore(), then dL/dIn = MultiLayerNetwork.epsilons()
i.e., see also this method and javadoc:
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L1178

**Comment by daanvdn** [comment]
Hi @AlexDBlack, 
So just be sure that I understand correctly: the methods you propose above would allow to grab the input parameters that a __previously trained model__ triggers for an __unseen__ example, right? 
If I want to use the "heat/saliency maps via gradients" approach, does the piece of code below make sense? 
```java 
MultiLayerNetwork trainedNetwork = ModelSerializer.restoreMultiLayerNetwork(new File("some_trained_model.bin")); 
        INDArray unseenExample = ...; 
        trainedNetwork.setInput(unseenExample); 
        INDArray allPossibleLabels = ...; 
        trainedNetwork.setLabels(allPossibleLabels); 
        trainedNetwork.computeGradientAndScore(); 
        INDArray triggeredInputParams = trainedNetwork.epsilon(); 
``` 
If so, I'm still not clear how I would be able to get hold of the activations in the outputLayer. I could be way off the mark, but could the following work to retrieve the predicted labels? 

```java  
INDArray predictedLabels = ((IOutputLayer) trainedNetwork.getOutputLayer()).getLabels(); 
``` 
Previously, when I was only interested in the predicted output labels, I would simply use `MultiLayerNetwork#feedForward(org.nd4j.linalg.api.ndarray.INDArray)`. This returns the activations for each layer, so you can just grab the last INDArray in the list and get the predicted labels from that. Obviously that makes no sense when I have already called `computeGradientAndScore`.


Thanks in advance for your feedback!

**Comment by AlexDBlack** [comment]
Yes, that code snippet is correct.
The .epsilon() array has identical shape to the unseenExample (input) array. Each entry: dL/da, where a[i] = input[i]
> If so, how do filter out those that have been triggered?

Not sure I totally understand what you are asking... but I *think* you are asking how to determine which ones are important? Well, you can look at the top 10% of absolute values, for example... or plot it as a heat map. There's no simple answer to that.

(Also I forgot to mention previously that with time series: you might want to use label masks to mask out all except the time step of interest)

> I'm not clear how I would be able to get hold of the predicted labels?
MultiLayerNetwork.output(input) is what you want. *technically* it's a little inefficient, but it's probably the easiest (in 99% of use cases you want the output and not the gradients - or vice/versa... your use case is an exception :))

**Comment by daanvdn** [comment]
great @AlexDBlack, that gives me a lot I can work with!

**Comment by AlexDBlack** [comment]
No problem. I'll close this issue as resolved.

**Comment by daanvdn** [comment]
Hi @AlexDBlack, a follow-up question re this issue and my code sample above.
Do I understand correctly that this way of collecting triggered input parameters for an unseen example behaves as if we are in training mode (even though we’re actually not training)? I’m asking because when I look at the implementation of ` org.deeplearning4j.nn.multilayer.MultiLayerNetwork#computeGradientAndScore` it says [here ](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L2135) that `train` is `true`. 
As far as the call to `MultiLayerNetwork#setLabels` in my code sample is concerned, it’s somewhat unclear to me what the INDArray that I pass should look like: 
-	Should it be the return value of ` MultiLayerNetwork.output(input)`? I.e. the labels with the weights that are predicted by the model? 
-	Or should I set the weights for every predicted label to 1.0?
-	Or should it be yet something else?

A final question I’m struggling with is the following. As far as can see, the code snippet above allows to grab the inputs that have been triggered for the __whole of the predicted labels__, right? But what if I want to know which inputs have been triggered for each __individual label__ (rather than all of them together)? If I want to achieve that, does it suffice to simply retain only the label of interest in the INDArray that I pass to `MultiLayerNetwork#setLabels`? Or will that not work?

Thanks again!

**Comment by shaulz** [comment]
Hi @daanvdn !

Can you please share the final code you have arrived in?

Thanks,
    Shaul
